### PR TITLE
Check the cluster ID (kube-system UID) when updating the cluster

### PR DESCRIPTION
**Pull request by yzxiu**
Signed-off-by: yzx <946666800@qq.com>

### What type of PR is this?
/kind feature

### What this PR does / why we need it:
When manual updating the cluster crd resource, check the cluster ID (kube-system UID)  to avoid heavy losses caused by copying wrong files.Because kubefed automatically performs a lot of irreversible operations

### Does this PR introduced a user-facing change?

```release-note
None
```

This also requires adding a validatingwebhookconfiguration,like
```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  annotations:
    meta.helm.sh/release-name: ks-core
    meta.helm.sh/release-namespace: kubesphere-system
  labels:
    app.kubernetes.io/managed-by: Helm
  name: cluster.kubesphere.io
webhooks:
- admissionReviewVersions:
  - v1beta1
  clientConfig:
    caBundle: <data>
    service:
      name: ks-controller-manager
      namespace: kubesphere-system
      path: /validate-cluster-kubesphere-io-v1alpha1
      port: 443
  failurePolicy: Fail
  matchPolicy: Exact
  name: validating-cluster.kubesphere.io
  namespaceSelector:
    matchExpressions:
    - key: control-plane
      operator: DoesNotExist
  objectSelector: {}
  rules:
  - apiGroups:
    - cluster.kubesphere.io
    apiVersions:
    - v1alpha1
    operations:
    - UPDATE
    resources:
    - clusters
    scope: '*'
  sideEffects: None
  timeoutSeconds: 30
```

I also submitted it in ks-installer https://github.com/kubesphere/ks-installer/pull/2131

### CODE REVIEW

File: cmd/controller-manager/app/server.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ package app
 import (
 	"context"
 	"fmt"
+	"kubesphere.io/kubesphere/pkg/controller/cluster"
```
Comment (iawia002) [inline cmd/controller-manager/app/server.go:22]: This should be moved to the next group.

**Comment by yzxiu** [comment]
@iawia002 sorry i accidentally closed https://github.com/kubesphere/kubesphere/pull/5293

File: pkg/controller/cluster/cluster_webhook.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"context"
+	"fmt"
+	v1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	clusterv1alpha1 "kubesphere.io/api/cluster/v1alpha1"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
```
Comment (iawia002) [inline pkg/controller/cluster/cluster_webhook.go:29]: This needs to be formatted like:

```go
	"context"
	"fmt"
	"net/http"

	v1 "k8s.io/api/admission/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/tools/clientcmd"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"

	clusterv1alpha1 "kubesphere.io/api/cluster/v1alpha1"
```

Comment (yzxiu) [reply pkg/controller/cluster/cluster_webhook.go]: formatted

File: cmd/controller-manager/app/server.go (hunk 1 above)
Comment (iawia002) [reply cmd/controller-manager/app/server.go:22]: @yzxiu You also need to update this

Comment (yzxiu) [reply cmd/controller-manager/app/server.go]: @iawia002 ok

File: pkg/controller/cluster/cluster_webhook.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	clusterConfig, err := clientcmd.RESTConfigFromKubeConfig(newCluster.Spec.Connection.KubeConfig)
+	if err != nil {
+		return admission.Denied(fmt.Sprintf("failed to load cluster config for %s: %s", newCluster.Name, err))
+	}
+
+	clusterClient, err := kubernetes.NewForConfig(clusterConfig)
```
Comment (iawia002) [inline pkg/controller/cluster/cluster_webhook.go:71]:
```suggestion
	clusterClient, err := kubernetes.NewForConfig(clusterConfig)
	if err != nil {
		return admission.Denied(err.Error())
	}
```

File: pkg/controller/cluster/cluster_webhook.go [hunk 4]
```diff
@@ -63,6 +64,10 @@ func (h *ValidatingHandler) Handle(ctx context.Context, req admission.Request) a
 		return admission.Errored(http.StatusBadRequest, err)
 	}
 
+	if reflect.DeepEqual(oldCluster.Spec, newCluster.Spec) {
+		return admission.Allowed("")
+	}
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/controller/cluster/cluster_webhook.go:69]: If only change the label or status, we don't need to check the UID every time.

**Comment by wansir** [comment]
/lgtm

@iawia002 any comments?

File: pkg/controller/cluster/cluster_webhook.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iawia002) [reply pkg/controller/cluster/cluster_webhook.go:69]: Nope, you can't tell if two clusters are the same by judging the `spec`, there is other information contained in the `spec`

**Pull request merged**